File: programmers/[programmers] 게임 맵 최단거리.py
from collections import deque
def solution(maps):
    answer = 0
    n, m = len(maps), len(maps[0])
    visited = [[0]*m for _ in range(n)]
    cnt = 1
    def bfs(y, x):
        nonlocal cnt, visited
        q = deque()
        q.append([y, x, cnt])
        direct = [[1, 0],[0 ,1],[-1, 0],[0, -1]]
        while q:
            y, x, cnt = q.popleft()
            if y==n-1 and x==m-1:
                return cnt
            for i in range(4):
                ny, nx = y+direct[i][0], x+direct[i][1]
                if ny>n-1 or nx>m-1 or ny<0 or nx<0: continue
                if visited[ny][nx]==1: continue
                if maps[ny][nx]==0: continue
                visited[ny][nx]=1
                q.append([ny, nx, cnt+1])
        return -1
    answer = bfs(0, 0)
    return answer

File: programmers/[programmers] 게임 맵 최단거리.js
function solution(maps) {
  const N = maps.length,
    M = maps[0].length;
  const BFS = (y, x) => {
    const d = [
      [1, 0],
      [-1, 0],
      [0, 1],
      [0, -1],
    ];
    const visited = Array.from({ length: N }, () => Array(M).fill(false));
    const q = [[y, x, 1]];
    visited[y][x] = true;

    while (q.length > 0) {
      const [y, x, cnt] = q.shift();
      if (y === N - 1 && x === M - 1) {
        return cnt;
      }
      for (let i = 0; i < 4; i++) {
        const ny = y + d[i][0];
        const nx = x + d[i][1];
        if (ny >= N || nx >= M || ny < 0 || nx < 0) continue;
        if (visited[ny][nx] === true) continue;
        if (maps[ny][nx] === 1) {
          visited[ny][nx] = true;
          q.push([ny, nx, cnt + 1]);
        }
      }
    }
    return -1;
  };
  return BFS(0, 0, 0);
}
